pkg/store/etcd.go

```diff
@@ -280,5 +293,117 @@ func (s *Etcd) DeleteTree(prefix string) error {
 // NewLock returns a handle to a lock struct which can be used to acquire and
 // release the mutex.
 func (s *Etcd) NewLock(key string, options *LockOptions) (Locker, error) {
-	return nil, ErrNotImplemented
+	var value string
+	ttl := uint64(time.Duration(defaultLockTTL).Seconds())
+
+	// Apply options
+	if options != nil {
+		if options.Value != nil {
+			value = string(options.Value)
+		}
+		if options.TTL != 0 {
+			ttl = uint64(options.TTL.Seconds())
+		}
+	}
+
+	// Create lock object
+	lock := &etcdLock{
+		client: s.client,
+		key:    key,
+		value:  value,
+		ttl:    ttl,
+	}
+
+	return lock, nil
+}
+
+// Lock attempts to acquire the lock and blocks while doing so.
+// Returns a channel that is closed if our lock is lost or an error.
+func (l *etcdLock) Lock() (<-chan struct{}, error) {
+
+	key := normalize(l.key)
+	var lastIndex uint64
+
+	// Lock holder channels
+	lockHeld := make(chan struct{})
+	stopLocking := make(chan struct{})
+
+	for {
+		resp, err := l.client.Create(key, l.value, l.ttl)
+		if err != nil {
+			if etcdError, ok := err.(*etcd.EtcdError); ok {
+				// Key already exists
+				if etcdError.ErrorCode != 105 {
+					lastIndex = ^uint64(0)
+				}
+			}
+		} else {
+			lastIndex = resp.Node.ModifiedIndex
+		}
+
+		_, err = l.client.CompareAndSwap(key, l.value, l.ttl, "", lastIndex)
+
+		if err == nil {
+			// Leader section
+			l.stopLock = stopLocking
+			go l.holdLock(key, lockHeld, stopLocking)
+			break
+		} else {
+			// Seeker section
+			chW := make(chan *etcd.Response)
+			chWStop := make(chan bool)
+			l.waitLock(key, chW, chWStop)
+
+			// Delete or Expire event occured
+			// Retry
+		}
+	}
+
+	return lockHeld, nil
+}
+
+// Hold the lock as long as we can
+// Updates the key ttl periodically until we receive
+// an explicit stop signal from the Unlock method
+func (l *etcdLock) holdLock(key string, lockHeld chan struct{}, stopLocking chan struct{}) {
+	defer close(lockHeld)
+
+	update := time.NewTicker(defaultUpdateTime)
+
+	for {
```

Comment:
@aluzzardi Well I attempted to do this but this is actually trickier because a race can occur between the `Update` and `CompareAndDelete` (with a wrong index) which will leave the cluster with no master until the key expires.

If we attempt the CAS multiple time to mitigate the race, processes not holding the lock are going to loop indefinitely because there is no way to succeed the CAS call and go out of the loop (unless saying we try the call 5 times and otherwise we break out of the loop, but this is a nasty workaround).

Same about watching the key, we are going to take actions on something that we probably did (let's say we deleted the key, we need to check wether it's us or another process that issued the action).

I'll leave this as it is with a FIXME, I already have a branch with those changes but I need to find a better solution to this issue :(